lib/experimental/class-wp-html-processor.php

```diff
@@ -0,0 +1,720 @@
+<?php
+/***
+ * Scans through an HTML document to find specific tags, then
+ * transforms those tags by adding, removing, or updating the
+ * values of the HTML attributes within that tag (opener).
+ *
+ * Does not fully parse HTML or _recurse_ into the HTML structure
+ * Instead this scans linearly through a document and only parses
+ * the HTML tag openers.
+ *
+ * @TODO:
+ *  - Write a comprehensive test suite for various unexpected
+ *    quirks and attribute names. Unfortunately we need to have
+ *    some fairly robust support because it's not difficult to
+ *    create misleading HTML through the block editor. For
+ *    example, as can happen with a number of plaintext inputs.
+ *
+ * Performance ideas:
+ *
+ * It may be possible that we're thrashing the garbage collector by
+ * allocating new classes for each tag and attribute token as we
+ * scan through the input document. If this is the case we could
+ * try pre-allocating a certain number of token class instances and
+ * reuse those upon scanning each new tag. This would incur a fixed
+ * memory cost but in addition to lowering the pressure on the
+ * garbage collector we might end up having more efficient use of
+ * localized data within the processor cache.
+ *
+ * If we can pick up scanning where we leave off, we can stop parsing
+ * within a tag opener as soon as we meet all the query constraints.
+ * If we need to set an attribute we haven't yet seen then we would
+ * need to continue scanning to the end of that tag opener until we
+ * find the existing attribute or know it's not there. This could
+ * avoid parsing large attributes, such as one with a long data-URI.
+ *
+ * Defer creating comparable values until we need them so we don't
+ * eagerly allocate and convert string values we don't ever read.
+ *
+ * Classes are passed by reference by default, but are we missing
+ * opportunities to pass by reference when we are copying by value?
+ */
+
+
+/**
+ * Processes an input HTML document by applying a specified set
+ * of patches to that input. Tokenizes HTML but does not fully
+ * parse the input document.
+ */
+class WP_HTML_Processor {
+	/**
+	 * Describes the constraints imposed on the HTML tag we're trying to match.
+	 *
+	 * @see WP_Tag_Find_Descriptor::parse()
+	 *
+	 * @var WP_Tag_Find_Descriptor
+	 */
+	public $descriptor;
+
+
+	/**
+	 * Scans the HTML document, keeping track of its own state.
+	 *
+	 * @var WP_HTML_Scanner
+	 */
+	public $scanner;
+
+
+	/**
+	 * Set of modifications to perform on input HTML document, built
+	 * via calls to the fluid API exposed by WP_HTML_Processor.
+	 *
+	 * @var WP_Tag_Modifications
+	 */
+	public $modifications;
+
+
+	public function __construct( $input ) {
+		$this->scanner = new WP_HTML_Scanner( $input );
+		$this->modifications = new WP_Tag_Modifications();
+	}
+
+
+	public function find( $query ) {
```

Comment:
I feel strongly about calling this `find_next`, otherwise the name suggests I could do this:

```php
$p = new WP_HTML_Processor('<div/><span/>');
$p->find(['tag_name'=>'span'])->set_attribute('id', 'span_id');
$p->find(['tag_name'=>'div'])->set_attribute('id', 'div_id');
```

Which is explicitly not supported. The documentation may explain it, but suggesting these nuances in the method names will save a lot of confusion.

Reply:
I'm confused because both of those finds will fail. Are you saying that to you, the name `find()` implies that each time it starts it will rewind to the beginning of the document?

Reply:
> Why would the first find fail? It finds the first span:

because in my head I read `</div></span>` and not what you wrote 🤦‍♂️

Reply:
> Yes, to me it does imply that.

Let's maybe not settle on this yet? I agree mostly with you, but also find `find()` quite reasonable for this purpose.

Reply:
`next()` sounds pretty good actually based on how we use it, and I don't think we need `_tag` in there especially if we end up with `tag` in the class name, which I think seems more and more appropriate as we go since this is so focused on tag openers alone